### ISSUE TITLE
sql: schema changes are too slow

### ISSUE THREAD

**Issue by vivekmenezes** [cockroachdb/cockroach#12424]
## Problem: Writing lots of keys -- like a large index -- is slow

creating an index over a 1M row block writer table (with traffic load) is taking around 17 minutes over a 5 node GCE cluster, which points to an index backfill insert rate of about 1000rows a second. Adding a column over a similar sized table happens in around 10 minutes. It would be nice if we can speed up both these by at least an order of magnitude. `SELECT COUNT(*)` over a 250K table takes 3 minutes so this is all to be expected.

Without using distributed-SQL we could speed this up by processing backfills in more goroutines, and better streaming the input data and the backfill output data.

**Comment by vivekmenezes** [comment]
With the move to creating an index via distsql the above index is taking around 10 minutes to create. While this is an improvement, more work needs to be done here.

**Comment by petermattis** [comment]
Agreed. I would expect each node to be able to handle in excess of 2000 rows/sec. If we can get all of the nodes performing inserts, I'd expect the index creation to take 1-2min.

**Comment by vivekmenezes** [comment]
Finally ran two schema changes on a table with 57M rows on a 6 node cluster:

```
I170404 15:52:59.098487 353 acceptance/allocator_test.go:234  starting schema change: CREATE INDEX foo ON datablocks.blocks (block_id)
I170404 15:52:59.098539 352 acceptance/allocator_test.go:234  starting schema change: ALTER TABLE datablocks.blocks ADD COLUMN newcol DECIMAL DEFAULT (DECIMAL '1.4')
I170404 18:45:11.738268 353 acceptance/allocator_test.go:239  completed schema change: CREATE INDEX foo ON datablocks.blocks (block_id), in 2h52m12.639734683s
I170404 20:24:54.685619 352 acceptance/allocator_test.go:239  completed schema change: ALTER TABLE datablocks.blocks ADD COLUMN newcol DECIMAL DEFAULT (DECIMAL '1.4'), in 4h31m55.587102319s
I170404 20:24:54.685663 22 acceptance/allocator_test.go:251  validate applied schema changes
I170404 21:54:10.779772 22 acceptance/allocator_test.go:277  SELECT COUNT(*) FROM datablocks.blocks AS OF SYSTEM TIME 1491337494768048234.0000000007: 57028870 rows
I170404 23:27:25.032594 22 acceptance/allocator_test.go:290  query: SELECT COUNT(newcol) FROM datablocks.blocks AS OF SYSTEM TIME 1491337494768048234.0000000007, found 57028870 rows
I170404 23:42:02.185115 22 acceptance/allocator_test.go:290  query: SELECT COUNT(block_id) FROM datablocks.blocks@foo AS OF SYSTEM TIME 1491337494768048234.0000000007, found 57028870 rows
```

The add index took around 2h50m, while the add column took around 1h40m. This acts as a great baseline to improve upon.

**Comment by petermattis** [comment]
57M index writes over 2h50m translates to ~5500 writes/sec. That's not too far off from what I would expect if the writes are random given the numbers we see on our test clusters. But the writes aren't random because you're adding the index `(block_id)` on a table that is sorted on `block_id`. Assuming you're doing batching for the index writes, I'd expect this to be an order of magnitude faster given the numbers we see for `BenchmarkSQL/Insert1/Cockroach` vs `BenchmarkSQL/Insert100/Cockroach`.

**Comment by petermattis** [comment]
`sql/distsqlrun/indexbackfiller.go` should be calling `client.Txn.CommitInBatch`, not `client.Run`. Failure to do this is will prevent the 1PC fast-path for transactions from being taken.

Also, `store.optimizePuts` only provides the blind-put optimization for `PutRequest` and `ConditionalPutRequest`. That's a valuable optimization, especially when there is a swath of empty key space that a batch of keys is being written into. Adding support for `InitPutRequest` should be easy. I'll file an issue about that.

**Comment by jordanlewis** [comment]
Thanks for the suggestion! Same goes for the columnbackfiller. I'll make a PR.

**Comment by petermattis** [comment]
> Thanks for the suggestion! Same goes for the columnbackfiller. I'll make a PR.

Great. Looking forward to seeing those schema change numbers improve.

**Comment by vivekmenezes** [comment]
```
 I170419 01:55:22.992277 16 acceptance/allocator_test.go:194  running schema changes while cluster is rebalancing
[01:55:22][Step 3/3] I170419 01:55:22.992351 334 acceptance/allocator_test.go:236  starting schema change: ALTER TABLE datablocks.blocks ADD COLUMN newcol DECIMAL DEFAULT (DECIMAL '1.4')
[01:55:22][Step 3/3] I170419 01:55:22.992795 335 acceptance/allocator_test.go:236  starting schema change: CREATE INDEX foo ON datablocks.blocks (block_id)
[03:28:23][Step 3/3] I170419 03:28:23.623615 334 acceptance/allocator_test.go:241  completed schema change: ALTER TABLE datablocks.blocks ADD COLUMN newcol DECIMAL DEFAULT (DECIMAL '1.4'), in 1h33m0.631196031s
[09:45:22][Step 3/3] I170419 09:45:22.492709 335 acceptance/allocator_test.go:241  completed schema change: CREATE INDEX foo ON datablocks.blocks (block_id), in 7h49m59.500358657s
[09:45:22][Step 3/3] I170419 09:45:22.492778 16 acceptance/allocator_test.go:253  validate applied schema changes
[11:36:32][Step 3/3] I170419 11:36:32.601825 16 acceptance/allocator_test.go:279  SELECT COUNT(*) FROM datablocks.blocks AS OF SYSTEM TIME 1492595122625823685.0000000000: 66764655 rows
[13:32:43][Step 3/3] I170419 13:32:43.316680 16 acceptance/allocator_test.go:292  query: SELECT COUNT(newcol) FROM datablocks.blocks AS OF SYSTEM TIME 1492595122625823685.0000000000, found 66764655 rows
[13:54:47][Step 3/3] I170419 13:54:47.885965 16 acceptance/allocator_test.go:292  query: SELECT COUNT(block_id) FROM datablocks.blocks@foo AS OF SYSTEM TIME 1492595122625823685.0000000000, found 66764655 rows
```

an index backfill on a 66M row table took 7h49s which is around 2345 ops/sec

**Comment by petermattis** [comment]
So the index backfill was half the speed this time? That's not good. Any idea what happened?

**Comment by cuongdo** [comment]
Did this test config get fixed with the new `COCKROACH_PREEMPTIVE_SNAPSHOT_RATE` setting? I don't maintain the schema change test, so I didn't modify it. The other allocator tests are using that.

I don't really see the connection, but it is a variable nonetheless...

**Comment by vivekmenezes** [comment]
@petermattis the original performance I had posted was incorrect because the index backfill ran before the column backfill in that test, and the number of rows statistic was computed after the column backfill was completed. FYI, blockwriter continues to write during both backfills. So in effect the original performance numbers were on a smaller table. The 2345 ops/sec number in the new results is more accurate.

These numbers are before jordan's CommitInBatch fix that he is working on.

@cuongdo thanks for pointing that out. I'll update the schema change tests to include that

**Comment by vivekmenezes** [comment]
`
[11:51:49][Step 3/3] I170517 11:51:49.414813 337 acceptance/allocator_test.go:236  starting schema change: CREATE INDEX foo ON datablocks.blocks (block_id)
[11:51:49][Step 3/3] I170517 11:51:49.414833 336 acceptance/allocator_test.go:236  starting schema change: ALTER TABLE datablocks.blocks ADD COLUMN newcol DECIMAL DEFAULT (DECIMAL '1.4')
[14:24:42][Step 3/3] I170517 14:24:42.645611 336 acceptance/allocator_test.go:241  completed schema change: ALTER TABLE datablocks.blocks ADD COLUMN newcol DECIMAL DEFAULT (DECIMAL '1.4'), in 2h32m53.230719548s
[04:51:30][Step 3/3] I170518 04:51:30.732406 337 acceptance/allocator_test.go:241  completed schema change: CREATE INDEX foo ON datablocks.blocks (block_id), in 16h59m41.317582771s
[04:51:30][Step 3/3] I170518 04:51:30.732450 20 acceptance/allocator_test.go:253  validate applied schema changes
[05:29:11][Step 3/3] I170518 05:29:11.496218 20 acceptance/allocator_test.go:279  SELECT COUNT(*) FROM datablocks.blocks AS OF SYSTEM TIME 1495083090848058607.0000000000: 57917878 rows
[06:07:51][Step 3/3] I170518 06:07:51.430735 20 acceptance/allocator_test.go:292  query: SELECT COUNT(newcol) FROM datablocks.blocks AS OF SYSTEM TIME 1495083090848058607.0000000000, found 57917878 rows
[06:16:20][Step 3/3] I170518 06:16:20.718111 20 acceptance/allocator_test.go:292  query: SELECT COUNT(block_id) FROM datablocks.blocks@foo AS OF SYSTEM TIME 1495083090848058607.0000000000, found 57917878 rows
`
Wow we took 16 hours to create an index over 57M rows.

**Comment by petermattis** [comment]
Have there been any efforts at figuring out where the time is going? It is nice to know that such a long schema change completes, but I'd suggest starting with a smaller schema change (e.g. a 100,000 row table) and tracking rows/second as the metric we want to improve.

**Comment by jordanlewis** [comment]
With #17078, we can do about 17k rows per second on an idle 4 node cluster with 100 ranges in the primary table. Note that for most of the time spent building the index, the index itself lives on a single range.

With modest load, we get an 8-10x slowdown. The `kv` flags used are:

```
  --concurrency 64
  --tolerate-errors
  --min-block-bytes=8
  --max-block-bytes=128
  --read-percent=50
  --max-rate=1200
```

I seems that the 8-10x slowdown comes from raft alone - most-to-all of the extra latency lives in Raft, as evidenced by the several-hundred-ms gap between the following log statements as viewed in an application of an `InitPut` batch in lightstep:
![image](https://user-images.githubusercontent.com/43821/28649263-879b2070-7241-11e7-80b3-6298f5268a2c.png)

Perhaps this slowdown is inevitable when the cluster is busy?

**Comment by vivekmenezes** [comment]
Nice!

It could be worth increasing the number of index entries we write per batch. With #17078 we have to way that against increasing the likelihood of a txn retry, because a large batch increases the chance of a retry because a single key update can retry the entire batch. Wanna see the effect of a large chunk size (2000 index entries). I should be better in an idle cluster but perhaps worse in an active one.

**Comment by tschottdorf** [comment]
Making the chunks larger will likely help somewhat (since it means less Raft cycles), but the ~220ms Raft applications hint at that there's a separate bottleneck here. @jordanlewis what latencies do the `kv` requests show? Are they also in the 220ms range? It would be really nice to have better visibility into Raft here.

**Comment by tschottdorf** [comment]
Not sure it'll help, but maybe this diff is good to have in the code anyway (probably more with threshold 10 or so):

```diff
diff --git a/pkg/storage/replica.go b/pkg/storage/replica.go
index 5c102519b..5b873946c 100644
--- a/pkg/storage/replica.go
+++ b/pkg/storage/replica.go
@@ -2716,6 +2716,14 @@ func (r *Replica) insertProposalLocked(
        if !proposal.Request.IsLeaseRequest() {
                r.mu.lastAssignedLeaseIndex++
        }
+
+       {
+               const inflightEventThreshold = 1
+               if diff := r.mu.lastAssignedLeaseIndex - r.mu.state.LeaseAppliedIndex; diff > inflightEventThreshold {
+                       log.Eventf(ctx, "%d commands inflight", diff)
+               }
+       }
+
        proposal.command.MaxLeaseIndex = r.mu.lastAssignedLeaseIndex
        proposal.command.ProposerReplica = proposerReplica
        proposal.command.ProposerLease = proposerLease
```

**Comment by dianasaur323** [comment]
@jordanlewis I believe we had decided to punt this to 1.3 so that you could work on SQL caching instead? I think 1.3 is a hard deadline for this, as well as some other schema change improvements though, so we can knock all of those out in one go. Moving to 1.3 unless you disagree.

**Comment by vivekmenezes** [comment]
I think this is fine although Jordan can start working earlier on this while 1.2 is being developed since he's already done a lot of thinking around this. That way it's rock solid for 1.3

**Comment by dianasaur323** [comment]
ok, that sounds like a solid plan to me. I believe that fixing schema changes is top on the backlog after caching sql plans, so we could optimistically get to it. Will keep in mind as things progress.

**Comment by vivekmenezes** [comment]
I'm planning on working on a prototype for index backfill using the new tech developed for csv load this week. I'll report on how far I get.

**Comment by vivekmenezes** [comment]
https://github.com/cockroachdb/cockroach/issues/19141

is related because when we implement the new schema change backfill it would be nice if it can backfill many indexes in parallel

**Comment by tomholub** [comment]
18 minutes here for `ALTER TABLE pubkey ADD COLUMN is_expired BOOLEAN NOT NULL DEFAULT FALSE;` on ~ 2M table, idle single-node roach. 

No index created - just a column - I too would have thought it would be faster, especially when it's a single node cluster.

**Comment by mholt** [comment]
I am seeing this too I think. For a 1-row table with about 15 columns, running:

```
ALTER TABLE "public"."accounts" ADD COLUMN "last_increase" TIMESTAMP;
```

Causes it to hang for several minutes. Quitting my client and trying to run it again yields:

```
ERROR:  duplicate: column "last_cloud_escrow_increase" in the middle of being added, not yet public
```

This is on a single-node cluster.

Screenshot from the admin panel - you can see this took 5 minutes:

<img width="1382" alt="screen shot 2018-09-25 at 9 00 04 pm" src="https://user-images.githubusercontent.com/1128849/46054794-1751c700-c106-11e8-9e5e-fe16cb34f61f.png">

**Comment by vivekmenezes** [comment]
@mholt what you're seeing seems very different. adding a column to a 1-row table should happen pretty rapidly. Are you able to reproduce this problem easily? If so I'd like to open another issue for your problem. Thanks for reporting this problem!

**Comment by mholt** [comment]
@vivekmenezes Yeah, I think so -- let me try again but we're in the middle of launching our product :) If someone beats me to it, cool -- if not, I'll try to post an issue within the week!

**Comment by dt** [comment]
@vivekmenezes I wrote up a summary of most of my investigations / conversations / prototype findings so far and put it in the issue desc. Now that I write it all in one place -- even after trying to break it up a bit with some sections and headings -- I'm thinking maybe I should have PRed it as a new section in the RFC or something.

**Comment by vivekmenezes** [comment]
@dt we discussed and writing a tech-note seems appropriate. Thanks!

**Comment by petermattis** [comment]
> creating an index over a 1M row block writer table (with traffic load) is taking around 17 minutes over a 5 node GCE cluster, which points to an index backfill insert rate of about 1000rows a second. Adding a column over a similar sized table happens in around 10 minutes. It would be nice if we can speed up both these by at least an order of magnitude. SELECT COUNT(*) over a 250K table takes 3 minutes so this is all to be expected.

@dt We can write data a lot faster than 1000 rows/sec on an idle cluster if it is batched appropriately. Can you elaborate on the `with traffic load`? Was this un-throttled load? Or was there some sort of throttling on the load so that the cluster wasn't already at its max capacity?

**Comment by dt** [comment]
I actually repurposed this issue and kept the original text as the first problem statement but I didn't actually write those two paragraphs, so I don't actually know what that was referring to specifically... @vivekmenezes might?.

That said there is some per-batch overhead and we currently do pretty small batches (I think 1k) since the whole batch needs to recreated if there's a conflict, so I'm not surprised if we're a bit slower that just batched inserts as fast as we can.

**Comment by vivekmenezes** [comment]
#32668

**Comment by vivekmenezes** [comment]
with the new backfill mechanism:

addindex: statement 1: "CREATE UNIQUE INDEX ON tpcc.order (o_entry_d, o_w_id, o_d_id, o_carrier_id, o_id);" took 38m4.430027s

this is over table tpcc.order with 30M rows

**Comment by vivekmenezes** [comment]
The performance of add sst is all over the place. This with it trying to add 5M index entries. perhaps it's not always adding 5M entries. Adding in more debug logging to check

```
logs/cockroach.log:I190201 04:03:17.938545 111819 sql/distsqlrun/indexbackfiller.go:153  [n1,client=71.125.15.117:56898,user=root,scExec,IndexBackfiller=60] index backfill stats: prepare 805.838052ms, sort 70.946265ms, add-sst 977.302887ms, flush 9.186722ms
logs/cockroach.log:I190201 04:04:11.372268 126926 sql/distsqlrun/indexbackfiller.go:153  [n1,client=71.125.15.117:56898,user=root,scExec,IndexBackfiller=60] index backfill stats: prepare 2.460451865s, sort 657.016499ms, add-sst 22.444239986s, flush 10.425117ms
logs/cockroach.log:I190201 04:04:51.292932 142779 sql/distsqlrun/indexbackfiller.go:153  [n1,client=71.125.15.117:56898,user=root,scExec,IndexBackfiller=60] index backfill stats: prepare 3.583515365s, sort 833.958439ms, add-sst 24.635835118s, flush 36.278788ms
logs/cockroach.log:I190201 04:05:48.976049 143342 sql/distsqlrun/indexbackfiller.go:153  [n1,client=71.125.15.117:56898,user=root,scExec,IndexBackfiller=60] index backfill stats: prepare 2.739245743s, sort 717.670749ms, add-sst 54.140441821s, flush 53.208358ms
logs/cockroach.log:I190201 04:08:03.501304 146755 sql/distsqlrun/indexbackfiller.go:153  [n1,client=71.125.15.117:56898,user=root,scExec,IndexBackfiller=60] index backfill stats: prepare 2.428667569s, sort 720.650881ms, add-sst 2m9.489395464s, flush 260.311077ms
logs/cockroach.log:I190201 04:16:09.233366 111759 sql/backfill.go:138  [n1,client=71.125.15.117:56898,user=root,scExec] Running backfill for "order", v=5, m=1
logs/cockroach.log:I190201 04:21:11.719938 242 sql/backfill.go:138  [n1,scExec] Running backfill for "order", v=5, m=1
logs/cockroach.log:I190201 04:21:26.309311 869830 sql/distsqlrun/indexbackfiller.go:153  [n1,IndexBackfiller=60] index backfill stats: prepare 15.723780303s, sort 1.306810109s, add-sst 10m3.543599879s, flush 681.883065ms
logs/cockroach.log:I190201 04:26:40.425419 3684147 sql/distsqlrun/indexbackfiller.go:153  [n1,IndexBackfiller=60] index backfill stats: prepare 522.573028ms, sort 1.514µs, add-sst 52.999µs, flush 1.223µs
logs/cockroach.log:I190201 04:27:00.733366 2745138 sql/distsqlrun/indexbackfiller.go:153  [n1,scExec,IndexBackfiller=60] index backfill stats: prepare 14.19814447s, sort 1.536411938s, add-sst 5m32.890879714s, flush 293.680489ms
logs/cockroach.log:W190201 04:27:00.746657 2745138 sql/distsqlrun/backfiller.go:285  [n1,scExec,IndexBackfiller=60] span /Table/60/1/{274/5/-2379/0-311/4/-1760} not found among [/Table/60/1/{477/4/-969/0-514/3/-368} /Table/60/1/5{51/3/-2742-95/8/-2025/0} /Table/60/{1/631/9/-987/0-2}]
logs/cockroach.log:I190201 04:32:01.813252 3940761 sql/distsqlrun/indexbackfiller.go:153  [n1,IndexBackfiller=60] index backfill stats: prepare 117.379468ms, sort 1.591µs, add-sst 61.737µs, flush 610ns
logs/cockroach.log:I190201 04:32:55.938639 3941133 sql/distsqlrun/indexbackfiller.go:153  [n1,IndexBackfiller=60] index backfill stats: prepare 94.188165ms, sort 1.913µs, add-sst 57.828µs, flush 757ns
logs/cockroach.vivek-1548990865-schemachange-bulkindex-tpcc-1000-0001.vivek.2019-02-01T03_15_43Z.003330.log:I190201 04:03:15.859063 111798 sql/event_log.go:135  [n1,client=71.125.15.117:56888,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:schemachanger.bulk_index_backfill.enabled Value:true User:root}
logs/cockroach.vivek-1548990865-schemachange-bulkindex-tpcc-1000-0001.vivek.2019-02-01T03_15_43Z.003330.log:I190201 04:03:16.069445 111759 sql/backfill.go:138  [n1,client=71.125.15.117:56898,user=root,scExec] Running backfill for "order", v=5, m=1
logs/cockroach.vivek-1548990865-schemachange-bulkindex-tpcc-1000-0001.vivek.2019-02-01T03_15_43Z.003330.log:I190201 04:03:17.938545 111819 sql/distsqlrun/indexbackfiller.go:153  [n1,client=71.125.15.117:56898,user=root,scExec,IndexBackfiller=60] index backfill stats: prepare 805.838052ms, sort 70.946265ms, add-sst 977.302887ms, flush 9.186722ms
logs/cockroach.vivek-1548990865-schemachange-bulkindex-tpcc-1000-0001.vivek.2019-02-01T03_15_43Z.003330.log:I190201 04:04:11.372268 126926 sql/distsqlrun/indexbackfiller.go:153  [n1,client=71.125.15.117:56898,user=root,scExec,IndexBackfiller=60] index backfill stats: prepare 2.460451865s, sort 657.016499ms, add-sst 22.444239986s, flush 10.425117ms
logs/cockroach.vivek-1548990865-schemachange-bulkindex-tpcc-1000-0001.vivek.2019-02-01T03_15_43Z.003330.log:I190201 04:04:51.292932 142779 sql/distsqlrun/indexbackfiller.go:153  [n1,client=71.125.15.117:56898,user=root,scExec,IndexBackfiller=60] index backfill stats: prepare 3.583515365s, sort 833.958439ms, add-sst 24.635835118s, flush 36.278788ms
logs/cockroach.vivek-1548990865-schemachange-bulkindex-tpcc-1000-0001.vivek.2019-02-01T03_15_43Z.003330.log:I190201 04:05:48.976049 143342 sql/distsqlrun/indexbackfiller.go:153  [n1,client=71.125.15.117:56898,user=root,scExec,IndexBackfiller=60] index backfill stats: prepare 2.739245743s, sort 717.670749ms, add-sst 54.140441821s, flush 53.208358ms
logs/cockroach.vivek-1548990865-schemachange-bulkindex-tpcc-1000-0001.vivek.2019-02-01T03_15_43Z.003330.log:I190201 04:08:03.501304 146755 sql/distsqlrun/indexbackfiller.go:153  [n1,client=71.125.15.117:56898,user=root,scExec,IndexBackfiller=60] index backfill stats: prepare 2.428667569s, sort 720.650881ms, add-sst 2m9.489395464s, flush 260.311077ms
logs/cockroach.vivek-1548990865-schemachange-bulkindex-tpcc-1000-0001.vivek.2019-02-01T03_15_43Z.003330.log:I190201 04:16:09.233366 111759 sql/backfill.go:138  [n1,client=71.125.15.117:56898,user=root,scExec] Running backfill for "order", v=5, m=1
logs/cockroach.vivek-1548990865-schemachange-bulkindex-tpcc-1000-0001.vivek.2019-02-01T03_15_43Z.003330.log:I190201 04:21:11.719938 242 sql/backfill.go:138  [n1,scExec] Running backfill for "order", v=5, m=1
logs/cockroach.vivek-1548990865-schemachange-bulkindex-tpcc-1000-0001.vivek.2019-02-01T03_15_43Z.003330.log:I190201 04:21:26.309311 869830 sql/distsqlrun/indexbackfiller.go:153  [n1,IndexBackfiller=60] index backfill stats: prepare 15.723780303s, sort 1.306810109s, add-sst 10m3.543599879s, flush 681.883065ms
logs/cockroach.vivek-1548990865-schemachange-bulkindex-tpcc-1000-0001.vivek.2019-02-01T03_15_43Z.003330.log:I190201 04:26:40.425419 3684147 sql/distsqlrun/indexbackfiller.go:153  [n1,IndexBackfiller=60] index backfill stats: prepare 522.573028ms, sort 1.514µs, add-sst 52.999µs, flush 1.223µs
logs/cockroach.vivek-1548990865-schemachange-bulkindex-tpcc-1000-0001.vivek.2019-02-01T03_15_43Z.003330.log:I190201 04:27:00.733366 2745138 sql/distsqlrun/indexbackfiller.go:153  [n1,scExec,IndexBackfiller=60] index backfill stats: prepare 14.19814447s, sort 1.536411938s, add-sst 5m32.890879714s, flush 293.680489ms
logs/cockroach.vivek-1548990865-schemachange-bulkindex-tpcc-1000-0001.vivek.2019-02-01T03_15_43Z.003330.log:W190201 04:27:00.746657 2745138 sql/distsqlrun/backfiller.go:285  [n1,scExec,IndexBackfiller=60] span /Table/60/1/{274/5/-2379/0-311/4/-1760} not found among [/Table/60/1/{477/4/-969/0-514/3/-368} /Table/60/1/5{51/3/-2742-95/8/-2025/0} /Table/60/{1/631/9/-987/0-2}]
logs/cockroach.vivek-1548990865-schemachange-bulkindex-tpcc-1000-0001.vivek.2019-02-01T03_15_43Z.003330.log:I190201 04:32:01.813252 3940761 sql/distsqlrun/indexbackfiller.go:153  [n1,IndexBackfiller=60] index backfill stats: prepare 117.379468ms, sort 1.591µs, add-sst 61.737µs, flush 610ns
logs/cockroach.vivek-1548990865-schemachange-bulkindex-tpcc-1000-0001.vivek.2019-02-01T03_15_43Z.003330.log:I190201 04:32:55.938639 3941133 sql/distsqlrun/indexbackfiller.go:153  [n1,IndexBackfiller=60] index backfill stats: prepare 94.188165ms, sort 1.913µs, add-sst 57.828µs, flush 757ns
```

**Comment by vivekmenezes** [comment]
spoke to david and we see

`adder, err := ib.flowCtx.BullkAdder(ctx, ib.flowCtx.ClientDB, 32<<10, readAsOf)`

it was meant to be

`adder, err := ib.flowCtx.BullkAdder(ctx, ib.flowCtx.ClientDB, 32<<20, readAsOf)`

it was flushing too aggressively.

Running along with that change.
